components/brave_wallet/common/f4_jumble.cc

```diff
@@ -0,0 +1,198 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/common/f4_jumble.h"
+
+#include <algorithm>
+#include <array>
+#include <memory>
+#include <string>
+
+#include "base/notreached.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "brave/third_party/argon2/src/src/blake2/blake2.h"
+
+namespace brave_wallet {
+
+const size_t kLeftSize = 64u;
+
+std::string GetHPersonalizer(uint8_t i) {
+  return {85,  65, 95,  70,  52, 74, 117,
+          109, 98, 108, 101, 95, 72, static_cast<char>(i),
+          0,   0};
+}
+
+std::string GetGPersonalizer(uint8_t i, uint16_t j) {
+  return {85,
+          65,
+          95,
+          70,
+          52,
+          74,
+          117,
+          109,
+          98,
+          108,
+          101,
+          95,
+          71,
+          static_cast<char>(i),
+          static_cast<char>(j & 0xFF),
+          static_cast<char>(j >> 8)};
+}
+
+std::optional<std::vector<uint8_t>> blake2b(const std::vector<uint8_t>& payload,
+                                            std::string_view personalizer,
+                                            size_t len) {
+  blake2b_state blake_state = {};
+  blake2b_param params = {};
+
+  if (personalizer.length() != sizeof(params.personal)) {
+    NOTREACHED();
+    return std::nullopt;
+  }
+
+  params.digest_length = len;
+  params.fanout = 1;
+  params.depth = 1;
+  memcpy(params.personal, personalizer.data(), sizeof(params.personal));
+  if (blake2b_init_param(&blake_state, &params) != 0) {
+    NOTREACHED();
+    return std::nullopt;
+  }
+  if (blake2b_update(&blake_state, payload.data(), payload.size()) != 0) {
+    NOTREACHED();
+    return std::nullopt;
+  }
+  std::vector<uint8_t> result(len);
+  if (blake2b_final(&blake_state, result.data(), len) != 0) {
+    NOTREACHED();
+    return std::nullopt;
+  }
+
+  return result;
+}
+
+std::vector<uint8_t> GetLeft(const std::vector<uint8_t>& message) {
+  size_t left_size =
+      std::min(static_cast<size_t>(kLeftSize), message.size() / 2);
+  return std::vector(message.begin(), message.begin() + left_size);
+}
+
+std::vector<uint8_t> GetRight(const std::vector<uint8_t>& message) {
+  size_t left_size =
+      std::min(static_cast<size_t>(kLeftSize), message.size() / 2);
+  return std::vector(message.begin() + left_size, message.end());
+}
+
+std::optional<std::vector<uint8_t>> h_round(uint8_t iter,
+                                            const std::vector<uint8_t>& left,
+                                            const std::vector<uint8_t>& right) {
+  if (left.size() == 0) {
+    return std::vector<uint8_t>();
+  }
+  auto hash = blake2b(right, GetHPersonalizer(iter), left.size());
+  if (!hash || hash->size() != left.size()) {
+    return std::nullopt;
+  }
+  std::vector<uint8_t> result;
```

Comment:
blake2b should never fail and always return hash size equal to left.size(). So this method never fails an no optional needed.
Same for g_round

Reply:
Done